### DISCUSSION TITLE
How to deploy Lowdefy v4 in an airgapped environment

### DISCUSSION THREAD

**Discussion by zinkls** [lowdefy/lowdefy#1645] [Questions and Help] [answered]
Hello, 

Unfortunately there is no docs for the v4 nodejs deployment yet.
I'm having trouble figuring out how to deploy my apps in a completely pre-built way.

Our internal apps made in lowdefy would be deployed in an airgapped environment - without outgoing network access.
Sadly I come from a .NET background and my experience with nextjs and advanced javascript build setups is very limited.

Currently, I can't use `lowdefy start` because there is no simple way to install npm-packages on the target system.
Is there any recommended way of build & deployment that allows me 
to simply run something like e.g. `node server.js` to host our lowdefy apps?

I'd be grateful for any advice in this matter :)

Version: v4.0.0-rc.9
Platform: nodejs on windows server

**Comment by SamTolmay** [comment, marked as answer]
Hi @zinkls 

I think if it is possible to use Docker that would be preferable, you can build the Docker image on another machine and run that image on your server. This will have a much lower chance of having any compatibility issues.

Else, you can follow the steps in the Docker file to build a standalone build of the app on another server.

The key steps you need to do are:
1. Set an environment `LOWDEFY_BUILD_OUTPUT_STANDALONE ` to `1` (This configures the build process to output he standalone build)
2. Run `pnpx lowdefy@rc build`. This will build the standalone app in the `<base-directory>/.lowdefy/server/.next/standalone` directory
3. Copy the contents of `<base-directory>/.lowdefy/server/.next/static` to `<base-directory>/.lowdefy/server/.next/standalone/.next/static` (Next.js prefers if these files are served by a CDN, but the server can serve them).
4. Copy the contents of `<base-directory>/public` to `<base-directory>/.lowdefy/server/.next/standalone/public`
5. The contents of the `<base-directory>/.lowdefy/server/.next/standalone` directory will now contain everything need to serve the app. This can be copied to your airgapped server and started using the command `node server.js`. 

You can read more about Next.js output standalone here: https://nextjs.org/docs/pages/api-reference/next-config-js/output

**Reply by zinkls** [reply]
Thank you :)
After a bit of testing it has worked out for me.
